src/shapes/instance.cpp

```diff
@@ -0,0 +1,201 @@
+        si.instance = this;
+
+        masked(si_out, active) = si;
+    }
+
+    std::pair<Vector3f, Vector3f> normal_derivative(const SurfaceInteraction3f &si,
+                                                    bool shading_frame,
+                                                    Mask active) const override {
+        MTS_MASK_ARGUMENT(active);
+
+        SurfaceInteraction3f temp(si);
+        temp.p     = m_to_object.transform_affine(si.p);
+        temp.dp_du = m_to_object.transform_affine(si.dp_du);
+        temp.dp_dv = m_to_object.transform_affine(si.dp_dv);
+
+        // Determine the length of the transformed normal before it was re-normalized
+        Normal3f tn = m_to_world.transform_affine(
+            normalize(m_to_object.transform_affine(si.sh_frame.n)));
+        Float inv_len = rcp(norm(tn));
+        tn *= inv_len;
+
+        auto [dn_du, dn_dv] = si.shape->normal_derivative(temp, shading_frame, active);
+
+        // Apply transform to dn_du and dn_dv
+        dn_du = m_to_world.transform_affine(Normal3f(dn_du)) * inv_len;
+        dn_dv = m_to_world.transform_affine(Normal3f(dn_dv)) * inv_len;
+
+        dn_du -= tn * dot(tn, dn_du);
+        dn_dv -= tn * dot(tn, dn_dv);
+
+        return { dn_du, dn_dv };
+    }
+
+    //! @}
+    // =============================================================
+
+#if defined(MTS_ENABLE_EMBREE)
+    RTCGeometry embree_geometry(RTCDevice device) const override {
+        if constexpr (!is_cuda_array_v<Float>) {
+            RTCGeometry instance = m_shapegroup->embree_geometry(device);
+            rtcSetGeometryTimeStepCount(instance, 1);
+            rtcSetGeometryTransform(instance, 0, RTC_FORMAT_FLOAT4X4_COLUMN_MAJOR, &m_to_world.matrix);
+            rtcCommitGeometry(instance);
+            return instance;
+        } else {
+            Throw("embree_geometry() should only be called in CPU mode.");
+        }
+    }
+#endif
+
+#if defined(MTS_ENABLE_OPTIX)
+    virtual void optix_prepare_instance(const OptixDeviceContext& context, OptixInstance& instance, uint32_t instance_id) override {
+        // TODO: check coeffs
+        float transform[12] = {
+            m_to_world.matrix(0, 0), m_to_world.matrix(0, 1), m_to_world.matrix(0, 2), m_to_world.matrix(0, 3),
+            m_to_world.matrix(1, 0), m_to_world.matrix(1, 1), m_to_world.matrix(1, 2), m_to_world.matrix(1, 3),
+            m_to_world.matrix(2, 0), m_to_world.matrix(2, 1), m_to_world.matrix(2, 2), m_to_world.matrix(2, 3)
+        };
+        std::memcpy(instance.transform, transform, sizeof(float) * 12);
+        instance.instanceId = instance_id;
+        instance.visibilityMask = 255;
+        instance.flags = OPTIX_INSTANCE_FLAG_NONE;
+        m_shapegroup->optix_accel_handle(context, instance.traversableHandle, instance.sbtOffset);
+    }
+
+    virtual void optix_fill_hitgroup_records(std::vector<HitGroupSbtRecord>&, OptixProgramGroup*) override {
+        /* no op */
+    }
```

Comment:
Do we need really need to override this method?

Reply:
Yes, because the base implementation appends a hitgroup, but an instance doesn't need any:
```
MTS_VARIANT void Shape<Float, Spectrum>::optix_fill_hitgroup_records(std::vector<HitGroupSbtRecord> &hitgroup_records, OptixProgramGroup *program_groups) {
    optix_prepare_geometry();
    // Set hitgroup record data
    hitgroup_records.push_back(HitGroupSbtRecord());
    hitgroup_records.back().data = { (uintptr_t) this, m_optix_data_ptr };

    size_t program_group_idx = (is_mesh() ? 2 : 3 + get_shape_descr_idx(this));
    // Setup the hitgroup record and copy it to the hitgroup records array
    rt_check(optixSbtRecordPackHeader(program_groups[program_group_idx], &hitgroup_records.back()));
}
```